src/main/java/com/google/step/FrontendQueryDatastore.java

```
@@ -142,18 +135,54 @@ private CompositeFilter buildCompositeFilter(
             compositeFilter = new CompositeFilter(CompositeFilterOperator.AND, filters);
         }
         return compositeFilter;
     }
 
+    /** Helper function for buildCityFilters **/
+    private Filter buildIndividualCityFilter(String city) {
+        return FilterOperator.EQUAL.of("City Name", city);
+    }
+
+    /** Builds the city filters for the overall Composite Filter **/
+    private Filter buildCityFilters(ArrayList<String> cityStrings) {
+        ArrayList<Filter> cityFilters = new ArrayList<>();
+        for (int i = 0; i < cityStrings.size(); i++) {
+            String city = cityStrings.get(i);
+            cityFilters.add(buildIndividualCityFilter(city));
+        }
+        if (cityFilters.size() > 1) {
+            return new CompositeFilter(CompositeFilterOperator.OR, cityFilters);
+        } else {
+            // We ensure that cityFilters is not empty in buildCompositeFilter().
+            return cityFilters.get(0);
+        }
+    }
+
+    /** Helper function for buildZoomFilters **/
+    private Filter buildIndividualZoomFilter(int zoom) {
+        return FilterOperator.EQUAL.of("Zoom", zoom);
+    }
+
     /** * Builds the zoom filters for the overall Composite Filter. * */
-    private Filter buildZoomFilters(int startingZoom, int endingZoom) {
+    private Filter buildZoomFilters(ArrayList<String> zoomStrings) {
         ArrayList<Filter> zoomFilters = new ArrayList<>();
-        for (int zoom = startingZoom; zoom <= endingZoom; zoom++) {
-            zoomFilters.add(FilterOperator.EQUAL.of("Zoom", zoom));
+        for (int i = 0; i < zoomStrings.size(); i++) {
+            try{
+                int zoom = Integer.parseInt(zoomStrings.get(i));
+                zoomFilters.add(buildIndividualZoomFilter(zoom));
+            } catch (NumberFormatException e) {
+                LOGGER.log(Level.WARNING, "Building Zoom Filters: " + e.getMessage());
+            }
```

Does this represent an actual error?  That is to say, something you were sent wasn't parsable?

If so, probably don't swallow the exception and move on.  Probably crashing on a bad request is reasonable behavior. (though of course, it's better to detect the problem, log something fatal, and then crash, rather than crashing without a log)

---

Essentially yeah, if something sent wasn't parsable the catch block would trigger. It wouldn't trigger if the zoom-levels array was empty as we check for that earlier. How do I make the program crash on a bad request?

---

You'll want to throw an exception!  In this case probably an IllegalArgumentException.

---

Sg!